ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/util/ScalablePushUtil.java

```diff
@@ -0,0 +1,109 @@
+ */
+
+package io.confluent.ksql.rest.util;
+
+import io.confluent.ksql.engine.KsqlEngine;
+import io.confluent.ksql.name.SourceName;
+import io.confluent.ksql.parser.OutputRefinement;
+import io.confluent.ksql.parser.tree.AliasedRelation;
+import io.confluent.ksql.parser.tree.AstVisitor;
+import io.confluent.ksql.parser.tree.Query;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.parser.tree.Table;
+import io.confluent.ksql.query.QueryId;
+import io.confluent.ksql.util.KsqlConfig;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+public final class ScalablePushUtil {
+
+  private static String STREAMS_AUTO_OFFSET_RESET_CONFIG = "auto.offset.reset";
+  private static String LATEST_VALUE = "latest";
+
+  private ScalablePushUtil() { }
+
+  @SuppressWarnings({"BooleanExpressionComplexity", "CyclomaticComplexity"})
+  public static boolean isScalablePushQuery(
+      final Statement statement,
+      final KsqlEngine ksqlEngine,
+      final KsqlConfig ksqlConfig,
+      final Map<String, Object> overrides
+  ) {
+    if (!ksqlConfig.getBoolean(KsqlConfig.KSQL_QUERY_PUSH_SCALABLE_ENABLED)) {
+      return false;
+    }
+    if (! (statement instanceof Query)) {
+      return false;
+    }
+    final Query query = (Query) statement;
+    final SourceFinder sourceFinder = new SourceFinder();
+    sourceFinder.process(query.getFrom(), null);
+    // It will be present if it's not a join, which we don't handle
+    if (!sourceFinder.getSourceName().isPresent()) {
+      return false;
+    }
+    // Find all of the writers to this particular source.
+    final SourceName sourceName = sourceFinder.getSourceName().get();
+    final Set<QueryId> queries = ksqlEngine.getQueriesWithSink(sourceName);
+    // See if the config or override have set the stream to be "latest"
+    final boolean isLatest = overrides.containsKey(STREAMS_AUTO_OFFSET_RESET_CONFIG)
+        ? LATEST_VALUE.equals(overrides.get(STREAMS_AUTO_OFFSET_RESET_CONFIG))
+        : LATEST_VALUE.equals(ksqlConfig.getKsqlStreamConfigProp(STREAMS_AUTO_OFFSET_RESET_CONFIG)
+            .orElse(null));
+    // Cannot be a pull query, i.e. must be a push
+    return !query.isPullQuery()
+        // Group by is not supported
+        && !query.getGroupBy().isPresent()
+        // Windowing is not supported
+        && !query.getWindow().isPresent()
+        // Having clause is not supported
+        && !query.getHaving().isPresent()
+        // Partition by is not supported
+        && !query.getPartitionBy().isPresent()
+        // There must be an EMIT CHANGES clause
+        && (query.getRefinement().isPresent()
+            && query.getRefinement().get().getOutputRefinement() == OutputRefinement.CHANGES)
+        // Must be reading from "latest"
+        && isLatest
+        // We only handle a single sink source at the moment from a CTAS/CSAS
+        && queries.size() == 1;
```

Comment:
Do we support the case if the upstream query is emitting a windowed table? For a windowed table, its `toStream` would result in a `KStream<Windowed<K>, V>` and the processor would be on `KeyValue<Windowed<K>, V>`. How is the key type being translated on the ksql layer?

Reply:
Yes, we do.  If you look in `ScalablePushRegistry.handleRow`, you can see that we take the key as an Object in the method, see if it's windowed, and if so, interprets the key as a `Windowed<GenericKey>`.

Reply:
kk, in that case if the push query filters on the key in `WHERE` clause, we'd need to first extract the `GenericKey` from `Windowed<GenericKey>` and then apply the operator right? Could you point me to the extraction logic just for my education? Tried to search it somewhere but did not found yet.